Support addition of gradient operations in a graph for golang

Cibifang opened the pull request:
related to  #21404

Cibifang committed:
Support addition of gradient operations in a graph for golang

asimshankar commented on tensorflow/go/graph.go:
```diff
@@ -147,6 +147,73 @@ func (g *Graph) Operations() []Operation {
+// i.e., d(y_1 + y_2 + ...)/dx_1, d(y_1 + y_2 + ...)/dx_2...
+// This is a variant of AddGradients that allows to caller to pass a custom
+// name prefix to the operations added to a graph to compute the gradients.
+func (g *Graph) AddGradientsWithPrefix(prefix string, y []Output, x []Output, dx []Output) ([]Output, error) {
```
Let's have just one method: `AddGradients` in Go (which can take a prefix string), instead of having multiple methods.

And then we can add a method to the `op` package, so that `Gradients` appears just like any other op (and can pick up the appropriate prefix from the `Scope`). A similar thing was done for the Java API.

That would allow common usage like:

```go
s := op.NewScope()
x := op.Placeholder(s, tf.Float) // Matrix to be provided to Session.Run
y := op.Mul(s, input, input)
dy := op.Gradients(s, []tf.Output{y}, []tf.Output{x})[0]
```

Cibifang replied:
Thanks for reviewing!

Some questions:
1. If the "appropriate prefix from the `Scope`” is `namespace` of `Scope`?
2. In the example, the method name is `Gradients`, is this a suggest name? Or I should still use `AddGradients`?
3. Also in the example, the `Gradients` Method only have three parameters(`scope`, `y[]`, `x[]`), don't have the `dx[]`, is this a suggestion that in the method the `dx[]` should just be set to `nil`?
4. I see all `op` method is in `wrappers.go`, same for this method?

An implementation example:
```
func Gradients(scope *Scope, y []tf.Output, x []tf.Output) (output [] tf.Output) {
	var err error
	if output, err = scope.graph.AddGradients(scope.namespace, y, x, nil); err != nil {
		scope.UpdateErr("Gradients", err)
		return
	}
	return output
}
```

asimshankar replied:
Thanks for the PR :)

1. Yup, see https://github.com/tensorflow/tensorflow/blob/7d3884bb87dc02c4548f55749f3d6db1b8364ddc/tensorflow/go/op/scope.go#L75

2. Let's use `Gradients`, as it's akin to a composite TensorFlow operation.

3. Perhaps we can treat `dx` as an optional parameter using the pattern we use for other operations. For example, see [`BatchMatMul`](https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go/op#BatchMatMul)

4. Nope, `wrappers.go` is automatically generated, I'd just add that to a new file `gradients.go` in the `op` directory.

Sound reasonable?

asimshankar reviewed (changes requested):
Thanks for the PR and sorry for the delay.
Some suggestions on integrating this better with the `op` package that is often more convenient for building graphs.

Cibifang committed:
Combine AddGradients and AddGradientsWithPrefix Methods in golang

Cibifang committed:
Fix gradients test in golang

1. fix some test name.
2. add test for existing prefix.

Cibifang committed:
Support addition of gradient operations in op package for golang

1. Add Gradients method in op package
  Enforce uniqueness of custom prefixes for gradients.
2. Add unit test for Gradients

Cibifang commented:
Hi, @asimshankar Can you take a look again?

Some details:
1. Different with Java implementation, `Gradients` have `prefix` as a parameters.
  If we just use `scope.namespace` as prefix, It will cause some problems:
    ```
    // When scope.namespace == ""
    // the grads1[0].Op.Name() will start with "gradients/"
    grads0 := Gradients(s,  []tf.Output{y0}, []tf.Output{x})

    // the grads1[0].Op.Name() will start with "gradients_1/"
    grads1 := Gradients(s,  []tf.Output{y0}, []tf.Output{x})
    ```
   It's OK because the scope.namespace is "", so the prefix is "".

   But something will be wrong when `scope.namespace is not ""
   ```
   // When scope.namespace == "sub"
    // the grads1[0].Op.Name() will start with "sub/"
    grads0 := Gradients(s,  []tf.Output{y0}, []tf.Output{x})

    // It will cause error because use conflicted prefix `sub`
    grads1 := Gradients(s,  []tf.Output{y0}, []tf.Output{x})
   ```

  In the Java Implementation, it is solved by a new class `NameScope` with new method `makeOpName`.
  ```
  // When scope.namespace == "sub"
  // If don't want to set prefix
  // grad0.dy(0).op().name() will start with "sub/Gradients/"
  Gradients grad0 = Gradients.create(scope, y, Arrays.asList(x));

  // If want to set prefix "MyGradients/"
  // grad1.dy(0).op().name() will start with "sub/MyGradients/`
  radients grad1 = Gradients.create(scope.withName("MyGradients"), y, Arrays.asList(x));
  ```

  to same with the implementation in Java, we need to add a `NameScope` struct, or add `opName`  to `Scope` structure(to be a default op name as prefix for Gradients). I don't know if it is good to do that for this feature. 
  So just add a `prefix` para to method `Gradients`

2. Enforce uniqueness of custom prefixes for gradients in method `Gradients` in op package, just like in Java

tensorflowbutler commented:
Nagging Assignee @aaroey: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

asimshankar commented on tensorflow/go/graph.go:
```diff
@@ -147,6 +147,66 @@ func (g *Graph) Operations() []Operation {
 	return ops
 }
 
+// AddGradientsWithPrefix adds operations to compute the partial derivatives of sum of `y`s w.r.t `x`s,
```
Update comment?

asimshankar commented on tensorflow/go/graph.go:
```diff
@@ -147,6 +147,66 @@ func (g *Graph) Operations() []Operation {
+		}
+	}
+
+	// If prefix is "", the C.TF_AddGradientsWithPrefix need cprefix to be nil but not ""
```
How about:

```go
if len(prefix) == 0 {
  cprefix = nil
}
C.TF_AddGradientsWithPrefix(...)
```

instead of repeating the larger call and all its arguments?

Cibifang replied:
It fixed. I just misunderstood the usage of `C.char` before.

asimshankar commented on tensorflow/go/op/gradients.go:
```diff
@@ -0,0 +1,41 @@
+//  dx: if not null, the partial derivatives of some loss function L w.r.t. y
+//
+//  return the partial derivatives
+func Gradients(scope *Scope, prefix string, y []tf.Output, x []tf.Output, dx ...tf.Output) (output []tf.Output) {
```
Can we avoid a `prefix` argument here altogether, that will be more consistent with other ops (where the name for the node in the graph isn't an argument)

Cibifang replied:
I'm not sure.
The meaning of `namespace` in `Scope` is not completely same with `prefix` for gradient.

For example:
If the `namespace` is `sub`, the  `Op.name()` of the `gradients` should be `sub/Gradients`;
But if we need a `sub/my_gradients`, there isn't other methods except for a `prefix` argument...

The JAVA implementation adds a `opName` to `Scope` structure(to be exact, it's a new `NameScope` structure). So they can add customized prefix by change change the value of `opName`.

If we actually don't want a `prefix` argument, we have two solutions (I see):
1. Always use `Gradients` as prefix, such as `Gradients`, `sub/Gradients`, `sub/Gradients_1`, etc

2. Add `opName` to `Scope` structure(also related methods like `SubScopeWithName`(to new a scope with customized opName) and `opName`(need to change behavior). It will be same with java implementation.

asimshankar reviewed (changes requested):
Thanks for the contributions, a few minor tweaks, but otherwise looks good.

Cibifang committed:
Fix comment in graph.go.

Method name is AddGradients but not AddGradientsWithPrefix.

Cibifang committed:
Fix calls to C.CString() in graph.go to avoid memory leaks.

Cibifang committed:
Remove  from func Gradients() in gradients.go

asimshankar commented on tensorflow/go/graph.go:
```diff
@@ -147,6 +147,67 @@ func (g *Graph) Operations() []Operation {
 
+// AddGradients adds operations to compute the partial derivatives of sum of `y`s w.r.t `x`s,
+// i.e., d(y_1 + y_2 + ...)/dx_1, d(y_1 + y_2 + ...)/dx_2...
+// This methods allows to caller to pass a custom name prefix to the operations
```
`// This method allows` instead of `// This methods allows`.

Or better just, how about just:

```go
// AddGradients add operations to compute the partial derivatives of the sum of tensors in y
// with respect to tensors in x, i.e., d(y[0] + y[1] + ...) / d x[0], d(y[0] + y[1] + ... ) / d x[1] etc.
//
// prefix, if non-empty, is the name prefix used for all operations added to the graph to compute
// these gradients.
```

Cibifang replied:
Fixed.

asimshankar commented on tensorflow/go/op/gradients.go:
```diff
@@ -0,0 +1,36 @@
+//  return the partial derivatives
+func Gradients(scope *Scope, y []tf.Output, x []tf.Output, dx ...tf.Output) (output []tf.Output) {
+	var err error
+	if output, err = scope.graph.AddGradients(scope.opName(scope.uniqueName("Gradients")), y, x, dx); err != nil {
```
Hmm...isn't this missing other elements of the `scope`? For example, control dependencies and device annotations (see https://github.com/tensorflow/tensorflow/blob/e45c90f0e4d17ac22048a73f1e81bd9c7a7a5145/tensorflow/go/op/scope.go#L85)

I'm okay if you'd want to be conservative and have this fall (by setting an error in `scope`) when `scope` had control dependencies or devices specified (that aren't handled here). Or better yet, handle them :)

Cibifang replied:
Sorry for the delay. I'm not sure what you are meaning about "missing other elements of the `scope`"...

If you are talking about `ControlDependencies` or `Device`, I think `WithControlDependencies` and `WithDevice` is already enough if who want to set them.

Of course we can also add it to `Gradient` and set it in `Gradient`,but the parameters will be more complex. 
Because the `dx` is already a optional parameters and it's a `[]Outputs`. So the declaration will be like below:
```
func Gradients(scope *Scope, y []tf.Output, x []tf.Output, option ...map[string]interface{}) (output []tf.Output)
```
And the call will be like:
```
sub := s.SubScope("sub")
grads1 := Gradients(sub, []tf.Output{y2}, {'dx':[]tf.Output{x1, x2}, 'device': '...', ''})
```
The implementation will be like:

asimshankar replied:
RE: "other elements of "scope", I believe the control dependencies and devices are _not_ being handled in the call to `Gradients(scope, ...)`.  

All the generated ops in the `op` package (like `Abs`, `Add` etc., see [`wrappers.go`](https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/go/op/wrappers.go)) invoke [`Scope.AddOperation`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/op/scope.go#L75), which in turns invokes [`Graph.AddOperation`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/graph.go#L191) which adds the control inputs and the device annotation via calls to `TF_AddControlInput` and `TF_SetDevice` etc.

However, the implementation of `Gradients` in this PR will not invoke those functions and thus will not respect any control dependencies or device annotations set in the `Scope` object.

Does that make sense? You could probably most easily see this by inspecting the `GraphDef` protocol buffer written out by `Graph.WriteTo()`.

I realize that fixing this would require some complexity (for example, the device annotation would probably have to be plumbed through the C API of `TF_AddGradients`, and the control dependencies may require wrapping the inputs in an `Identity` operation or something). So, for this PR, I'm okay if we don't handle them but instead have `Gradients()` fail if it sees non-empty control dependencies or a device in the `Scope`. Does that make sense?

Cibifang replied:
Thanks for explanation but I am sorry that I still have some confusion.

According the implementation and test case in C/C++ and Java. `AddGradient` is not like with other op of scope.

For example, we can see test case in C++, [TEST_F(GradientsTest, TwoMatMuls_Chained)](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients_test.cc#L139-L170).
In this case, we can see the common op like `Const`/`MatMul` and gradient op `AddSymbolicGradients`. 
In [here](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/c/c_api_test.cc#L1695-L1710), These common op is implemented with `TF_NewOperation` and `TF_FinishOperation`, so they can invoke `TF_AddInput` or `TF_AddControlInput` or `TF_SetDevice`(All these functions need a `TF_OperationDescription` returned by `TF_NewOperation`.
But [`AddSymbolicGradients`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients.cc#L557-L565) doesn't invoke `TF_NewOperation`, so it can't invoke `TF_AddControlInput` or `TF_SetDevice` because it don't have a `TF_OperationDescription`.

Our implementation in Go actually needs to invode C implementation. So I'm confused how to invoke `TF_AddControlInput` or `TF_SetDevice` when the c implementation doesn't invoke these function...
If we just use [`WithControlDependencies`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/op/scope.go#L115-L133), it will be effective until next "common op" is used.

Cibifang replied:
In fact, [the implementation of C](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/c/c_api.cc#L2441-L2447) uses [the api about scope in C++](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients.h#L30-L35).

And the scope of C++ also have [device](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/scope.h#L132-L134) and [controlDepandencies](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/scope.h#L119-L126)
But the implementation of  gradient don't handle them.

I'm not sure if it is right to implement it with different logic in C++(when we finally call the implementation in C++).

Of course, if we still think it it more important to be consistent with other op in go or there are other details that I didn't find, I'm also willing to just lead to an error in the code now.
@asimshankar

Cibifang replied:
@asimshankar Can you take a look again?

asimshankar commented on tensorflow/go/op/gradients.go:
```diff
@@ -0,0 +1,36 @@
+//  return the partial derivatives
+func Gradients(scope *Scope, y []tf.Output, x []tf.Output, dx ...tf.Output) (output []tf.Output) {
+	var err error
+	if output, err = scope.graph.AddGradients(scope.opName(scope.uniqueName("Gradients")), y, x, dx); err != nil {
```
I'm not sure we want to use `scope.uniqueName` here, just `scope.opName("Gradients")`.
The reason I say so is that we don't use this for any other ops (https://github.com/tensorflow/tensorflow/blob/e45c90f0e4d17ac22048a73f1e81bd9c7a7a5145/tensorflow/go/op/scope.go#L85 etc.), so I wouldn't do so here either - to be consistent with how primitive operations are added to the graph.

Cibifang replied:
Fixed.

Cibifang committed:
Fix comment for AddGradients in graph.go

Cibifang committed:
Stop using uniqueName for Gradients in gradient.go

To keep the code style consistent with other functions.

tensorflowbutler commented:
It has been 20 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

Cibifang commented:
sorry, I'll update it soon.

asimshankar reviewed (changes requested):
@Cibifang : RE:  https://github.com/tensorflow/tensorflow/pull/21895#discussion_r234851776 , yeah, let's just error out for now in case there are fields in scope that are set and not being respected.

Cibifang committed:
Fix comment for AddGradients in graph.go

Cibifang committed:
Raise error if trying to use Gradients when scope have device or
control dependencies specified(Gradients doesn't currently support them).

Cibifang commented on tensorflow/go/graph.go:
```diff
@@ -147,6 +147,68 @@ func (g *Graph) Operations() []Operation {
 	return ops
 }
 
+// AddGradients add operations to compute the partial derivatives of the sum of tensors in y
```
fixed
